### DISCUSSION TITLE
Custom disk driver for "Cloudinary"

### DISCUSSION THREAD

**Discussion by iF-ckAdonis** [adonisjs/core#3659] [Help]
I'm struggling to write custom disk driver for [Cloudinary](https://cloudinary.com). In the doc, it says that every driver must implement [`DriverContract`](https://github.com/adonisjs/drive/blob/develop/adonis-typings/drive.ts#L53-L134) interface, but some of the APIs are "technically incompatible" with the upstream vendor. For example, `put` method has been implemented to not return anything, but [`cloudinary` SDK](https://cloudinary.com/documentation/image_upload_api_reference#upload_response) returns the data of the uploaded image/video. Another thing: what should `visibility` be of the images/videos hosted in **Cloudinary**? Is it considered with respect to **Cloudinary** itself or to the application?

**Comment by thetutlage** [comment]
> For example, put method has been implemented to not return anything, but [cloudinary SDK](https://cloudinary.com/documentation/image_upload_api_reference#upload_response) returns the data of the uploaded image/video

You can just ignore the data returned by cloudinary and do not return it from your implementation.

> Another thing: what should visibility be of the images/videos hosted in Cloudinary? Is it considered with respect to Cloudinary itself or to the application?

This is the privacy of the uploaded file on Cloudinary itself. If `visibility` is public, then anyone with the file URL can access it. Otherwise a signed URL is required

**Reply by thetutlage** [reply]
The primary goal of Drive is to make it easy to switch between drivers without vendor lock-in. For example.

Let's say the Cloudinary driver returns an object from the `PUT` method and your code reads properties from that object. But now, during development you want to use the `local` driver and during testing you want to use the memory driver. Both the `local` and the `memory` drivers does not return back this object, therefore you will have to write all sort conditionals to handle these scenarios. 

If you think you want to use Cloudinary all the time. Then it will be better and easier to use their Node SDK directly vs wrapping it as a Drive driver

**Comment by Saykot262** [comment]
<https://github.com/adonisjs/bodyparser/blob/2bc5652fc7291ef10cef806cfa48bb10891fe7d6/src/Multipart/File.ts#L238> and <https://github.com/adonisjs/bodyparser/blob/2bc5652fc7291ef10cef806cfa48bb10891fe7d6/src/Multipart/File.ts#L248> took 2 days of my coworker @msrumon to finally figure out to not use `moveToDisk()` method of the uploaded file while using **Cloudinary**.

To the maintainers: please make it more flexible and allow us to use customized "key" of ourselves.
